File: packages/mastra-adapter/src/tool-registry.ts
export type ToolExecutionFunction = (
  executionContext: any
) => Promise<any> | any;

export interface ToolRegistryEntry {
  execute: ToolExecutionFunction;
}

export class MastraToolRegistry {
  private tools: Map<string, ToolRegistryEntry> = new Map();

  /**
   * Register a tool with its execution function
   * @param name - The tool name (must match the tool name in MDX)
   * @param entry - The tool registry entry with execute function
   */
  register(name: string, entry: ToolRegistryEntry): void {
    this.tools.set(name, entry);
  }

  /**
   * Register multiple tools at once
   * @param toolEntries - Object with tool names as keys and registry entries as values
   */
  registerTools(toolEntries: Record<string, ToolRegistryEntry>): void {
    Object.entries(toolEntries).forEach(([name, entry]) => {
      this.register(name, entry);
    });
  }

  /**
   * Get a tool's registry entry by name
   * @param name - The tool name
   * @returns The tool registry entry if found
   * @throws Error if tool not found
   */
  get(name: string): ToolRegistryEntry {
    const tool = this.tools.get(name);
    if (!tool) {
      throw new Error(`Tool not found: ${name}`);
    }
    return tool;
  }

  /**
   * Check if a tool is registered
   * @param name - The tool name
   * @returns true if the tool is registered
   */
  has(name: string): boolean {
    return this.tools.has(name);
  }

  /**
   * Get all registered tool names
   * @returns Array of registered tool names
   */
  getToolNames(): string[] {
    return Array.from(this.tools.keys());
  }

  /**
   * Remove a tool from the registry
   * @param name - The tool name to remove
   * @returns true if the tool was removed, false if it didn't exist
   */
  unregister(name: string): boolean {
    return this.tools.delete(name);
  }

  /**
   * Clear all registered tools
   */
  clear(): void {
    this.tools.clear();
  }

  /**
   * Get the number of registered tools
   * @returns The number of registered tools
   */
  size(): number {
    return this.tools.size;
  }
}

File: packages/pydantic-ai-v0-adapter/src/agentmark_pydantic_ai_v0/tool_registry.py
"""Type-safe tool registry for Pydantic AI adapter."""

from __future__ import annotations

from .types import AsyncToolFunction, RegisteredTool, ToolFunction


class PydanticAIToolRegistry:
    """Tool registry matching TypeScript VercelAIToolRegistry patterns.

    Tools are registered with their execution functions. When the adapter
    builds Pydantic AI Tool instances, it looks up the execution function
    from this registry.

    Example:
        registry = PydanticAIToolRegistry()

        # Register sync tool
        registry.register("search", lambda args, ctx: search_web(args["query"]))

        # Register async tool
        async def fetch_data(args: dict, ctx: dict | None) -> dict:
            return await api.get(args["url"])
        registry.register("fetch", fetch_data)

        # Register tool that needs Pydantic AI's RunContext
        registry.register("db_query", db_tool_fn, takes_ctx=True)
    """

    def __init__(self) -> None:
        """Initialize empty tool registry."""
        self._tools: dict[str, RegisteredTool] = {}

    def register(
        self,
        name: str,
        fn: ToolFunction | AsyncToolFunction,
        takes_ctx: bool = False,
    ) -> PydanticAIToolRegistry:
        """Register a tool execution function.

        Args:
            name: Tool name (must match AgentMark config tool key).
            fn: Execution function with signature (args, tool_context) -> result.
                The tool_context is passed from AdaptOptions.toolContext.
            takes_ctx: Whether the tool needs Pydantic AI's RunContext
                       in addition to the AgentMark tool_context.

        Returns:
            Self for method chaining.
        """
        self._tools[name] = RegisteredTool(
            name=name,
            execute=fn,
            takes_ctx=takes_ctx,
        )
        return self

    def get(self, name: str) -> RegisteredTool | None:
        """Get a registered tool by name.

        Args:
            name: Tool name.

        Returns:
            RegisteredTool or None if not found.
        """
        return self._tools.get(name)

    def has(self, name: str) -> bool:
        """Check if a tool is registered.

        Args:
            name: Tool name.

        Returns:
            True if tool is registered.
        """
        return name in self._tools

    def list_names(self) -> list[str]:
        """List all registered tool names.

        Returns:
            List of tool names.
        """
        return list(self._tools.keys())

    def remove(self, name: str) -> bool:
        """Remove a tool from the registry.

        Args:
            name: Tool name.

        Returns:
            True if tool was removed, False if not found.
        """
        if name in self._tools:
            del self._tools[name]
            return True
        return False
